source/src/microapp/cs_MicroappStorage.cpp

```diff
@@ -173,27 +173,69 @@ uint16_t MicroappStorage::validateChunk(const uint8_t * const data, uint16_t siz
 }
 
 /**
- * This function validates the app in flash. It can only be used when the state is written and all flash operations
- * have finished. Only use it when all chunks, also the last chunk has been written successfully.
+ * This function validates the app in flash. Only use it when all chunks, also the last chunk has been written 
+ * successfully.
  *
  * The checksum is calculated iteratively, by using the chunk size. For the last chunk we will only use the part 
  * up to the total size of the binary. By doing it iteratively we can keep the local buffer relatively small.
  */
 uint16_t MicroappStorage::validateApp() {
 	TYPIFY(STATE_MICROAPP) state_microapp;
+
+	LOGd("Validate app");
+	uint16_t ret_code;
+
 	cs_state_id_t id = 0;
 	cs_state_data_t data(CS_TYPE::STATE_MICROAPP, id, (uint8_t*)&state_microapp, sizeof(state_microapp));
 	State::getInstance().get(data);
+
+	LOGd("Found microapp of size: %x", state_microapp.size);
 	
+	LOGd("Compare with checksum: %x", state_microapp.checksum);
 	uint8_t buf[MICROAPP_CHUNK_SIZE];
+	uint32_t addr;
+
+	microapp_storage.start_addr = g_FLASH_MICROAPP_BASE;
+	
+	addr = microapp_storage.start_addr;
+
+	nrf_fstorage_read(&microapp_storage, addr, &buf, 16);
+	int offset = (int)((buf[1] << 8) + (buf[0]));
+	LOGd("Offset: %i", offset);
+	int size = (int)((buf[5] << 8) + (buf[4]));
+	LOGd("Size: %i", size);
+
+	if (size == 0xFFFF) {
+		LOGw("Do not load");
+		return ERR_WRONG_STATE;
+	}
+	int g_checksum = (int)((buf[9] << 8) + (buf[8]));
+	LOGd("Checksum: %x", g_checksum);
+
+	state_microapp.start_addr = microapp_storage.start_addr;
+	state_microapp.offset = offset;
+	state_microapp.size = size;
+	state_microapp.checksum = g_checksum;
+
+	LOGd("Write microapp config");
+	State::getInstance().set(data);
 
 	// read from flash with fstorage, calculate checksum
 	uint8_t count = state_microapp.size / MICROAPP_CHUNK_SIZE;
 	uint16_t remain = state_microapp.size - (count * MICROAPP_CHUNK_SIZE);
 	uint32_t checksum_iterative = 0;
-	uint32_t addr = microapp_storage.start_addr;
+	
 	for (int i = 0; i < count; ++i) {
-		nrf_fstorage_read(&microapp_storage, addr, &buf, MICROAPP_CHUNK_SIZE);
+		ret_code = nrf_fstorage_read(&microapp_storage, addr, &buf, MICROAPP_CHUNK_SIZE);
+		if (ret_code != ERR_SUCCESS) {
+			LOGw("Error with reading with fstorage: %i", ret_code);
+			return ERR_INVALID_MESSAGE;
+		}
+		if (i == 0) {
+			for (int j = 0; j < 16; ++j) {
+				buf[j] = 0;
+			}
+		}
```

Comment:
What's this? Why not use memset? And where does 16 come from?

Reply:
Fixed in https://github.com/crownstone/bluenet/commit/cf20c97dfe77cce02d2a80555f843335b0e7fa04.

Reply:
The magic constant 16 is gone. It is rewritten such that it is clear that there is a checksum over the header first and then over the binary. For the Fletcher checksum the first is basically a no-op because it should return 0 on all zeros. However, if we go to CRC16, we'll need it.